### PR TITLE
fix: Behavior of CLI when no arguments are passed

### LINKED ISSUE #14308: Omiting any file/dir positional parameter to CLI does not give an error
* **ESLint Version:** 7.23.0
* **Node Version:** 15.4.0
* **npm Version:** 7.1.1
* **Operating System:** linux 5.4.0-70-generic
* **Parser:** @babel/eslint-parser

<details>
<summary>Configuration</summary>

```js
module.exports = {
  root: true,
  parserOptions: {
    parser: '@babel/eslint-parser',
    ecmaVersion: 2018,
    sourceType: 'module'
  },
  extends: [
    'standard',
    'plugin:vue/recommended'
  ],
  plugins: ['vue'],
  rules: {
    'max-len': ['warn', { code: 120 }],
    'vue/no-v-html': 'off'
  }
}
```

</details>

ESLint CLI usage:

```bash
eslint --cache --ext .js,.vue --max-warnings 0
```

**What did you expect to happen?**

I expected to have and error in console: `'patterns' must be a non-empty string or an array of non-empty strings`.

Instead, I got no response at all (and no actual linting performed, of course). 

```
$ eslint --cache --ext .js,.vue --max-warnings 0
Done in 0.23s.
```

**How to replicate?**

1. Run `eslint` CLI with no positional parameters. 
2. No error is shown .

**More details**

The [condition in ESLint engine code](https://github.com/eslint/eslint/blob/71a80e38aab2dada01b808ed43d9b0e806d863c4/lib/eslint/eslint.js#L523) is accepting an empty list as a valid list of non empty strings. It boils down to the fact that a call to `.every(...)` method of an empty list is always true ([see `isArrayOfNonEmptyString` function](https://github.com/eslint/eslint/blob/71a80e38aab2dada01b808ed43d9b0e806d863c4/lib/eslint/eslint.js#L119))

I'm more than OK to prepare a PR to fix this.

Thanks!

### PR DESCRIPTION
**Pull request by nzakas**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[x] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

When running the CLI without file arguments it silently fails right now. This PR makes the following changes:

* When running the CLI in eslintrc mode, failing to use file arguments throws an error.
* When running the CLI in flat config mode, failing to use file arguments means that it runs on the current working directory.

Fixes #14308

#### Is there anything you'd like reviewers to focus on?

To make this work, I changed how `lintFiles()` works in both `ESLint` and `FlatESLint`. In `FlatESLint`, I decided that an empty string and an empty array are the same and should be treated as `.`. I'm not sure if that's the right decision, so I'd like to hear some opinions on that.

### CODE REVIEW

**Comment by mdjermanovic** [comment]
> I decided that an empty string and an empty array are the same and should be treated as `.`. I'm not sure if that's the right decision, so I'd like to hear some opinions on that.

Makes sense to me.

File: lib/eslint/flat-eslint.js [hunk 1]
```diff
@@ -731,7 +731,21 @@ class FlatESLint {
...
+         *
+         * In both cases, we use the cwd as the directory to lint.
+         */
+        if (patterns === "" || Array.isArray(patterns) && patterns.length === 0) {
+            normalizedPatterns = ["."];
+        } else if (typeof patterns === "string") {
+            normalizedPatterns = [patterns];
+        } else if (!isArrayOfNonEmptyString(patterns)) {
             throw new Error("'patterns' must be a non-empty string or an array of non-empty strings");
         }
```
Comment (mdjermanovic) [inline lib/eslint/flat-eslint.js:750]: Maybe something like this so that strings with only spaces like `" "` are caught by the validation as it used to be before this change:

```suggestion
        } else {
            if (!isNonEmptyString(patterns) && !isArrayOfNonEmptyString(patterns)) {
                throw new Error("'patterns' must be a non-empty string or an array of non-empty strings");
            }

            if (typeof patterns === "string") {
                normalizedPatterns = [patterns];
            }
        }
```

File: lib/eslint/eslint-helpers.js [hunk 2]
```diff
@@ -105,20 +105,20 @@ class AllFilesIgnoredError extends Error {
...
  * Check if a given value is an array of non-empty strings or not.
- * @param {any} x The value to check.
- * @returns {boolean} `true` if `x` is an array of non-empty strings.
+ * @param {any} value The value to check.
+ * @returns {boolean} `true` if `value` is an array of non-empty strings.
  */
-function isArrayOfNonEmptyString(x) {
-    return Array.isArray(x) && x.every(isNonEmptyString);
+function isArrayOfNonEmptyString(value) {
+    return Array.isArray(value) && value.length && value.every(isNonEmptyString);
```
Comment (mdjermanovic) [inline lib/eslint/eslint-helpers.js:121]: This function is also used to validate `ignorePatterns` constructor option, so after this change passing `ignorePatterns: []` to the constructor will become invalid. Is that intentional?

Comment (nzakas) [reply lib/eslint/eslint-helpers.js:121]: Oops no. I thought I caught all the side effects of this change, but apparently missed this one.

**Comment by fasttime** [comment]
Let me point out a use case that I think will be negatively impacted by this change. Sometimes people run ESLint only on a subset of files that is calculated dynamically and that could be possibly empty. For example, for performance reasons, some users will want to lint only changed files in a git repo, so they could be running a command like this ([reference](https://stackoverflow.com/a/63842983/1083663)):

```sh
eslint --fix $(git diff --name-only HEAD | xargs)
```

After this PR, if a user runs the above command while there are no uncommitted changes, the whole current directory will be linted (or they will get an error if eslintrc is used).

**Comment by fasttime** [comment]
> I assume that ESLint isn't the only tool people are using in this way. There are a lot of tools that default to `.` when no file arguments are present. How do other tools handle this use case?

I'd guess that people are using ESLint like this simply because they were able do it so far. `npx prettier` prints usage info and exits with an error code. The same happens when I run `javac` (the Java compiler binary) without arguments. On the other hand, `ls` without arguments just lists the contents of the current directory, so probably there is no standard way to handle this edge case.

**Comment by nzakas** [comment]
I think `jest` works in such a way that you can either pass no arguments or specific tests? Maybe there's some way to set that up for a precommit hook that we can learn from? (I'm not familiar with Jest, so looking for some insights.)

**Comment by fasttime** [comment]
In fact running `jest` without arguments prints an error message and exits with an error code:

```console
PS D:\foo> npx jest                  
No tests found, exiting with code 1
Run with `--passWithNoTests` to exit with code 0
In D:\foo
  8 files checked.
  testMatch: **/__tests__/**/*.[jt]s?(x), **/?(*.)+(spec|test).[tj]s?(x) - 0 matches
  testPathIgnorePatterns: \\node_modules\\ - 8 matches
  testRegex:  - 0 matches
Pattern:  - 0 matches

```

But this behavior can be changed with a flag ([`--passwithnotests`](https://jestjs.io/docs/cli#--passwithnotests)):

```console
PS D:\foo> npx jest --passWithNoTests
No tests found, exiting with code 0
```

**Comment by fasttime** [comment]
> Maybe there's some way to set that up for a precommit hook that we can learn from?

It seems that `lint-staged` does not run tasks for glob patterns that don't match any staged file, so at least it shouldn't be a concern for this change.

**Comment by nzakas** [comment]
Jest is failing without tests because it didn't find any in the directories that it searches by default. It seems like `--passwithnotests` might be the best thing to copy.

**Comment by sam3k** [comment]
> @sam3k please be sure to leave comments as to the resolution from the TSC meeting along with removing the "tsc agenda" label.

Sounds good

**Comment by nzakas** [comment]
@mdjermanovic this is ready for review

File: tests/lib/cli.js [hunk 3]
```diff
@@ -815,6 +815,15 @@ describe("cli", () => {
                         assert.strictEqual(exit, useFlatConfig ? 0 : 2);
                     });
 
+                    it(`should not lint anything when no files are passed if --pass-on-no-patterns is passed with configType:${configType}`, async () => {
+                        const exit = await cli.execute("--pass-on-no-patterns", null, useFlatConfig);
+
+                        assert.isFalse(log.info.called);
+
+                        // When eslintrc is used, we get an exit code of 2 because the --no-warn-ignored option is unrecognized.
```
Comment (mdjermanovic) [inline tests/lib/cli.js:823]:
```suggestion
```

I think this comment doesn't apply here,

File: tests/lib/eslint/eslint.js [hunk 4]
```diff
@@ -1537,6 +1572,19 @@ describe("ESLint", () => {
             }, /All files matched by 'tests\/fixtures\/\*-quoted\.js' are ignored\./u);
         });
 
+        it("should not throw an error when ignorePatterns is an empty array", async () => {
+            eslint = new ESLint({
+                useEslintrc: false,
+                overrideConfig: {
+                    ignorePatterns: []
+                }
```
[the remark this answers is not in the document]
Comment (nzakas) [reply tests/lib/eslint/eslint.js:1582]: I had to change the behavior of option validation, so I added this test in to make sure that the existing behavior was maintained.

File: lib/eslint/flat-eslint.js [hunk 5]
```diff
@@ -727,20 +729,50 @@ class FlatESLint {
...
      * Executes the current configuration on an array of file and directory names.
-     * @param {string|string[]} patterns An array of file and directory names.
+     * @param {string|string[]} [patterns=["."]] An array of file and directory names.
      * @returns {Promise<LintResult[]>} The results of linting the file patterns given.
      */
-    async lintFiles(patterns) {
-        if (!isNonEmptyString(patterns) && !isArrayOfNonEmptyString(patterns)) {
-            throw new Error("'patterns' must be a non-empty string or an array of non-empty strings");
-        }
+    async lintFiles(patterns = ["."]) {
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/eslint/flat-eslint.js:735]: Hmm, that's a good point. I thought API-wise it might make sense to match the CLI behavior when no arguments are passed, but I can see how this can be confusing.

File: docs/src/use/command-line-interface.md [hunk 6]
```diff
@@ -716,6 +726,18 @@ npx eslint --exit-on-fatal-error file.js
 npx eslint --no-warn-ignored --max-warnings 0 ignored-file.js
 ```
 
+#### `--pass-on-no-patterns`
+
+This option allows ESLint to exit with code 0 when no file or directory patterns are passed. Without this option, ESLint assume you wan to use `.` as the pattern. (When running in legacy eslintrc mode, ESLint will exit with code 1.)
```
Comment (mdjermanovic) [inline docs/src/use/command-line-interface.md:731]:
```suggestion
This option allows ESLint to exit with code 0 when no file or directory patterns are passed. Without this option, ESLint assumes you want to use `.` as the pattern. (When running in legacy eslintrc mode, ESLint will exit with code 1.)
```

Just noticed two typos.

**Comment by nzakas** [comment]
Argh, totally messed up the rebase on this one. :(

**Comment by nzakas** [comment]
I believe this is ready.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!
